File: OOPRecords.Model/SubjectReport.cs
using NakedObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace OOPRecords.Model
{
    public class SubjectReport 
    {
        public TeacherRepository TeacherRepository { set; protected get; }
        public SubjectRepository SubjectRepository { set; protected get; }

        [Hidden(WhenTo.Always)]
        public virtual int Id { get; set; }

        [MemberOrder(1)][Disabled]
        public virtual Student Student { get; set; }

        [MemberOrder(2)]
        public virtual Subject Subject { get; set; }

        public IQueryable<Subject> AutoCompleteSubject(string match)
        {
            return SubjectRepository.FindSubjectByName(match);
        }

        [MemberOrder(3)]
        public virtual string Grade { get; set; }

        public IList<string> ChoicesGrade()
        {
            return new List<string> { "A*", "A", "B", "C", "D", "E", "U" };
        }

        [MemberOrder(4)]
        public virtual Teacher GivenBy { get; set; }

        public IList<Teacher> ChoicesGivenBy()
        {
            return TeacherRepository.AllTeachers().ToList();
        }

        [MemberOrder(5)]
        [Mask("d")]
        public virtual DateTime Date { get; set; }

        public DateTime DefaultDate()
        {
            return DateTime.Today;
        }

        [MemberOrder(6)][MultiLine][Optionally]
        public virtual string Notes { get; set; }

        public override string ToString()
        {
            return $"{Subject}, {Date}";
        }
    }
}

File: OOPRecords.ModelVB/SubjectReport.vb
Imports NakedObjects

Public Class SubjectReport

    Private _teacherRep As TeacherRepository
    Public WriteOnly Property TeacherRepository() As TeacherRepository
        Set(ByVal value As TeacherRepository)
            _teacherRep = value
        End Set
    End Property

    Private _subjectRep As SubjectRepository
    Public WriteOnly Property SubjectRepository() As SubjectRepository
        Set(ByVal value As SubjectRepository)
            _subjectRep = value
        End Set
    End Property


    <Hidden(WhenTo.Always)>
    Public Overridable Property Id As Integer

    <MemberOrder(1)> <Disabled>
    Public Overridable Property Student As Student

    <MemberOrder(2)>
    Public Overridable Property Subject As Subject

    Public Function AutoCompleteSubject(ByVal match As String) As IQueryable(Of Subject)
        Return _subjectRep.FindSubjectByName(match)
    End Function

    <MemberOrder(3)>
    Public Overridable Property Grade As String

    Public Function ChoicesGrade() As IList(Of String)
        Return New List(Of String) From {"A*", "A", "B", "C", "D", "E", "U"}
    End Function

    <MemberOrder(4)>
    Public Overridable Property GivenBy As Teacher

    Public Function ChoicesGivenBy() As IList(Of Teacher)
        Return _teacherRep.AllTeachers().ToList()
    End Function

    <MemberOrder(5)>
    <Mask("d")>
    Public Overridable Property OnDate As Date

    Public Function DefaultDate() As Date
        Return Date.Today
    End Function

    <MemberOrder(6)> <MultiLine> <Optionally>
    Public Overridable Property Notes As String

    Public Overrides Function ToString() As String
        Return $"{Subject}, {OnDate}"
    End Function
End Class